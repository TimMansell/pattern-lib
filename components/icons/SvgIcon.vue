<template>
  <svg :width="width" :height="height" focusable="false" role="presentation">
    <use :xlink:href="'#' + name" />
  </svg>
</template>

<script>
export default {
  props: ['name', 'width', 'height'],
};
</script>
