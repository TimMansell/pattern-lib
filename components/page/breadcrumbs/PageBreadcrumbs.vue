<template>
  <nav aria-label="Breadcrumb">
    <ol class="page-breadcrumbs max links-reset">
      <li
        v-for="(item, index) in items"
        :key="item.href"
        class="page-breadcrumbs__item"
      >
        <a class="page-breadcrumbs__link" :href="item.href" :aria-current="index === items.length -1 ? 'page' : false">
          <SvgIcon
            v-if="index === 0"
            class="page-breadcrumbs__home"
            name="home"
            width="18" height="18"
          />
          <span class="page-breadcrumbs__name">{{ item.text }}</span>
          <SvgIcon
            v-if="index < items.length - 1"
            class="page-breadcrumbs__chevron"
            name="chevron-right"
            width="12" height="12"
          />
        </a>
      </li>
    </ol>
  </nav>
</template>

<script>
import SvgIcon from '../../icons/SvgIcon.vue';

export default {
  components: { SvgIcon },
  props: {
    items: {
      type: Array,
      required: true,
      validator: arr => (
        arr.length > 0 &&
        arr.every(el => !!el.href && !!el.text)
      ),
    },
  },
};
</script>
