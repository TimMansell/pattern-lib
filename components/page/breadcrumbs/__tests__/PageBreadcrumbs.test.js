import { mount } from 'vue-test-utils';
import SvgIcon from '../../../icons/SvgIcon.vue';
import PageBreadcrumbs from '../PageBreadcrumbs.vue';

describe('PageBreadcrumbs', () => {
  describe('`items` prop validator', () => {
    const { validator } = PageBreadcrumbs.props.items;

    it('rejects empty array', () => {
      expect(validator([])).toBeFalsy();
    });

    it('rejects items missing `href` and/or `text`', () => {
      expect(validator([{}])).toBeFalsy();
      expect(validator([{ href: '/foo' }])).toBeFalsy();
      expect(validator([{ text: 'Foo' }])).toBeFalsy();
    });

    it('accepts items with non-empty `href` and `text`', () => {
      expect(validator([{ href: '/foo', text: 'Foo' }])).toBeTruthy();
    });
  });

  describe('with two items', () => {
    let cmp;

    beforeEach(() => {
      cmp = mount(PageBreadcrumbs, {
        propsData: {
          items: [
            { href: '/foo', text: 'Foo' },
            { href: '/bar', text: 'Bar' },
          ],
        },
      });
    });

    it('should render an array of breadcrumb items', () => {
      const elems = cmp.findAll('li');
      expect(elems.length).toBe(2);
      expect(elems.at(0).find('a').hasAttribute('href', '/foo')).toBe(true);
      expect(elems.at(0).text()).toContain('Foo');
      expect(elems.at(1).find('a').hasAttribute('href', '/bar')).toBe(true);
      expect(elems.at(1).text()).toContain('Bar');
    });

    it('should indicate the position of each item to search engines', () => {
      const elems = cmp.findAll('meta[itemprop="position"]');
      expect(elems.length).toBe(2);
      expect(elems.at(0).hasAttribute('content', '1')).toBe(true);
      expect(elems.at(1).hasAttribute('content', '2')).toBe(true);
    });

    it('should render a `home` and a `chevron-right` icon inside the first item', () => {
      const elems = cmp.find('li').findAll(SvgIcon);
      expect(elems.at(0).hasProp('name', 'home')).toBe(true);
      expect(elems.at(1).hasProp('name', 'chevron-right')).toBe(true);

      // Ensure the icons are actually rendered (cf. below)
      expect(elems.at(0).html()).toContain('svg');
      expect(elems.at(1).html()).toContain('svg');
    });

    it('should not render any icon inside the last item', () => {
      // Seems like `vue-test-utils` is dumb when it comes to `v-if`...
      // `findAll(SvgIcon)` finds the icons even though they don't get rendered.
      const html = cmp.findAll('li').at(1).html();
      expect(html).not.toContain('svg');
    });

    it('should match snapshot', () => {
      expect(cmp.element).toMatchSnapshot();
    });
  });
});
