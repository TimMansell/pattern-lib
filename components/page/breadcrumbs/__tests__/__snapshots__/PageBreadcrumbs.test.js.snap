// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBreadcrumbs with two items should match snapshot 1`] = `
<nav
  aria-label="Breadcrumb"
>
  <ol
    class="page-breadcrumbs max links-reset"
  >
    <li
      class="page-breadcrumbs__item"
    >
      <a
        class="page-breadcrumbs__link"
        href="/foo"
      >
        <svg
          class="page-breadcrumbs__home"
          focusable="false"
          height="18"
          role="presentation"
          width="18"
        >
          <use
            xlink:href="#icon-home"
          />
        </svg>
         
        <span
          class="page-breadcrumbs__name"
        >
          Foo
        </span>
         
        <svg
          class="page-breadcrumbs__chevron"
          focusable="false"
          height="12"
          role="presentation"
          width="12"
        >
          <use
            xlink:href="#icon-chevron-right"
          />
        </svg>
      </a>
    </li>
    <li
      class="page-breadcrumbs__item"
    >
      <a
        aria-current="page"
        class="page-breadcrumbs__link"
        href="/bar"
      >
        <!---->
         
        <span
          class="page-breadcrumbs__name"
        >
          Bar
        </span>
         
        <!---->
      </a>
    </li>
  </ol>
</nav>
`;
